src/Corax/Analyzers/Analyzer.cs

```diff
@@ -0,0 +1,281 @@
+            where TTokenizer : ITokenizer
+            where TTransform1 : ITransformer
+        {
+            return Create<TTokenizer, TTransform1, NullTransformer, NullTransformer>(tokenizer, transform1);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Analyzer Create<TTokenizer, TTransform1, TTransform2>(in TTokenizer tokenizer = default(TTokenizer),
+                            in TTransform1 transform1 = default(TTransform1),
+                            in TTransform2 transform2 = default(TTransform2))
+            where TTokenizer : ITokenizer
+            where TTransform1 : ITransformer
+            where TTransform2 : ITransformer
+        {
+            return Create<TTokenizer, TTransform1, TTransform2, NullTransformer>(tokenizer, transform1, transform2);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Analyzer With<TTransform1>(in TTransform1 transform1 = default(TTransform1))
+            where TTransform1 : ITransformer
+        {
+            return new Analyzer(this, &Run<NullTokenizer, TTransform1, NullTransformer, NullTransformer>,
+                default(NullTokenizer), new ITransformer[] { transform1 });
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Analyzer With<TTransform1, TTransform2>(
+                            in TTransform1 transform1 = default(TTransform1),
+                            in TTransform2 transform2 = default(TTransform2))
+            where TTransform1 : ITransformer
+            where TTransform2 : ITransformer
+        {
+            return new Analyzer(this, &Run<NullTokenizer, TTransform1, TTransform2, NullTransformer>,
+                default(NullTokenizer), new ITransformer[] { transform1, transform2 });
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Analyzer With<TTransform1, TTransform2, TTransform3>(in TTransform1 transform1 = default(TTransform1), 
+                            in TTransform2 transform2 = default(TTransform2), 
+                            in TTransform3 transform3 = default(TTransform3))
+            where TTransform1 : ITransformer
+            where TTransform2 : ITransformer
+            where TTransform3 : ITransformer
+        {
+            return new Analyzer(this, &Run<NullTokenizer, TTransform1, TTransform2, TTransform3>,
+                default(NullTokenizer), new ITransformer[] { transform1, transform2, transform3 });
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Execute(ReadOnlySpan<byte> source, ref Span<byte> output, ref Span<Token> tokens)
+        {
+            Debug.Assert(tokens.Length >= (int)(_sourceBufferMultiplier * source.Length));
+            Debug.Assert(output.Length >= (int)(_tokenBufferMultiplier * source.Length));
+
+            _func(this, source, ref output, ref tokens);
```

Comment:
What guarantees you have that the `source` wouldn't produce more than can be fit into `tokens`?

Reply:
It is the transformation responsibility to do that calculation. For example ngram will generate more than source, but you know that `_tokenBufferMultiplier` for the ngram tokenizer has to be at least `n` so what you are controlling there is that they are bigger. However, I noticed there that `token` and `output` are inverted.

Reply:
This looks like a stumbling block waiting to happen.

Reply:
Building proper analyzing blocks would require serious engineering anyways. Not such a big deal.